### PR TITLE
Hybrid detector with LLM fallback

### PR DESCRIPTION
**Pull request by pradh**
Usage:
* The detector is controlled via the url param `detector` which takes: `heuristic`, `llm`, `hybrid` (new).
* Default is still `heuristic`.  Next week, plan to switch over to `hybrid` after more testing.
 
Impl:
* The main logic is in `llm_fallback.use_llm()`.  Rest of the PR is util code shuffling (to `detection` subdir) and param plumbing.

Screenshots:

image
image

### CODE REVIEW

File: server/lib/nl/detection/llm_fallback.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+from server.lib.nl.common import counters
+from server.lib.nl.detection import utils
+from server.lib.nl.detection.types import ClassificationType
+from server.lib.nl.detection.types import Detection
+from shared.lib import detected_variables as vars
+
+_COMPLEX_QUERY_TOKEN_THRESHOLD = 8
+
+
+#
```
Comment (shifucun) [inline server/lib/nl/detection/llm_fallback.py:27]: For ease of future updates, good to list all the "llm" query examples in the function comment.

File: server/lib/nl/detection/utils.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  top_multi_sv_score = multi_sv.candidates[0].aggregate_score
+
+  # Prefer multi-sv when the scores are higher or up to a score differential.
+  if (top_multi_sv_score > top_sv_score or top_sv_score - top_multi_sv_score
+      <= shared_constants.MULTI_SV_SCORE_DIFFERENTIAL):
+    return True
+  return False
+
+
+def has_dual_sv(detection: Detection) -> bool:
```
Comment (chejennifer) [inline server/lib/nl/detection/utils.py:73]: add a function comment about what the function does?

File: server/lib/nl/detection/utils.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  if (not single_sv or not single_sv.scores or not multi_sv or
+      not multi_sv.candidates):
+    return False
+  # Get the top single-sv and multi-sv scores.
+  top_sv_score = single_sv.scores[0]
+  top_multi_sv_score = multi_sv.candidates[0].aggregate_score
+
+  # Prefer multi-sv when the scores are higher or up to a score differential.
+  if (top_multi_sv_score > top_sv_score or top_sv_score - top_multi_sv_score
+      <= shared_constants.MULTI_SV_SCORE_DIFFERENTIAL):
```
Comment (chejennifer) [inline server/lib/nl/detection/utils.py:68]: nit: could this if condition just be `top_multi_sv_score - top_sv_score >= -shared_constants.MULTI_SV_SCORE_DIFFERENTIAL`

**Review by chejennifer — COMMENTED** [review]
cool!

File: server/lib/nl/detection/llm_fallback.py (hunk 1 above)
Comment (pradh) [reply server/lib/nl/detection/llm_fallback.py:27]: Done.

File: server/lib/nl/detection/utils.py (hunk 3 above)
Comment (pradh) [reply server/lib/nl/detection/utils.py:68]: I slightly prefer the readability of the current version, than x - y >= -10 :)

File: server/lib/nl/detection/utils.py (hunk 2 above)
Comment (pradh) [reply server/lib/nl/detection/utils.py:74]: Done.

File: server/routes/nl/api.py [hunk 4]
```diff
@@ -100,18 +105,20 @@ def data():
         uttr_history=escaped_context_history,
         debug_counters=counters.get(),
         query_detection_debug_logs=query_detection_debug_logs,
-        use_llm=False)
+        detector='Heuristic Based')
```
[the remark this answers is not in the document]
Comment (pradh) [reply server/routes/nl/api.py:110]: The types to names aren't 1:1.  Added constants for detector types.  There isn't much repetition in names, and they are temporary, so leaving those as strings.

File: static/js/apps/nl_interface/query_result.tsx [hunk 5]
```diff
@@ -73,16 +73,17 @@ export const QueryResult = memo(function QueryResult(
...
+    if (props.detector !== null) {
+      detectorStr = props.detector;
     }
-
     axios
-      .post(`/api/nl/data?q=${query}&idx=${indexType}${useLLMStr}`, {
-        contextHistory: props.contextHistory,
-      })
+      .post(
+        `/api/nl/data?q=${query}&idx=${indexType}&detector=${detectorStr}`,
```
[the remark this answers is not in the document]
Comment (pradh) [reply static/js/apps/nl_interface/query_result.tsx]: Good catch.  I did the same for indexType too.

**Review by pradh — COMMENTED** [review]
Thanks for the comments!

**Comment by pradh** [comment]
> Some of the lint errors worth fix https://app.codacy.com/gh/datacommonsorg/website/pullRequest?prid=12052271

Tried to address all the medium ones... will take another look in this round.

**Review by chejennifer — APPROVED** [review]
thanks for the updates!

**Review by shifucun — APPROVED** [review]

**Pull request merged**